### PR TITLE
Fixing objectosphere bug

### PR DESCRIPTION
**Pull request by akshay-raj-dhamija**
Addresses issue #45

### CODE REVIEW

**Comment by siebenkopf** [comment]
This bug fix does not address the real problem. The problem is in line 148 (https://github.com/Vastlab/vast/pull/47/files#diff-50abf7dc1d80a62d5bb80c32503b87f00caeccab9d4b66fdd57b89aa0f436744R148), i.e., when you `detach` from the automatic gradient computation. You should not do that since otherwise

**Comment by akshay-raj-dhamija** [comment]
Yes, just pushed the actual fix and now there is a difference in the results.
This is the result of new training, now the validation accuracy for MNIST has gone down to ~96% from ~99% while the training accuracy remains the same and the histogram also shows the difference.
Would have to try some different parameters to check if it can be trained better
letters_2D_plot
Histogram
[letters_hist.pdf](https://github.com/Vastlab/vast/files/7554624/letters_hist.pdf)

**Comment by siebenkopf** [comment]
I see. Thanks for fixing.

I believe that we have seen this behavior before, i.e., that not all the feature space was taken over by the network. This is, hopefully, only an artifact that is based on random initialization. I am running again to see if this happens to me as well.

**Comment by siebenkopf** [comment]
I have run the code using:
```
python MNIST_SoftMax_Training.py --approach objectosphere --dataset_root \Source\data --no_of_epochs 100 --solver adam --Minimum_Knowns_Magnitude 5
```

and for me everything looks normal:

letters_2D_plot
[letters_hist.pdf](https://github.com/Vastlab/vast/files/7568576/letters_hist.pdf)

**Review by siebenkopf — APPROVED** [review]

File: vast/losses/losses.py [hunk 1]
```diff
@@ -1,7 +1,6 @@
 import torch
 from torch.nn import functional as F
-
-from .. import tools
+from vast import tools
```
[the remark this answers is not in the document]
Comment (siebenkopf) [reply vast/losses/losses.py]: I have undone this.

**Review by siebenkopf — APPROVED** [review]

File: vast/losses/losses.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/losses/losses.py]: Thank you @siebenkopf 
The change was made thinking it was much more readable and it was working for me.
Just so that I understand this correctly, `from vast import tools` would fail only if `vast` has not been installed with `pip` or not in your `PYTHONPATH`?
I can see another corner case problem here in case multiple versions of `vast` are present in which case this may import the version installed rather than the version in the current working directory.
I see many other repos like the official pytorch repo taking the approach `from torch.ao.quantization.fake_quantize import FakeQuantizeBase `.
Is my understanding of the above problem correct, because I think I am doing this at many other places in the codebase, which might need to be reverted?

[the remark this answers is not in the document]
Comment (siebenkopf) [reply vast/losses/losses.py]: Generally: the package is defined in the `__init__.py` files in the directories. Only what you import there can be accessed trough your main package. For example, if you have the following structure:
```
package
- __init__.py
- code.py
```

with contents:
```
code.py:
def some_func():
  pass
```

```
__init__.py:
from . import code
```

Then you can call:
```
>>> import package
>>> package.code.some_func()
```

If you would leave the `__init__.py` empty, you would need to call:
``` 
>>> import package.code
>>> package.code.some_func()
```

Note that the `code.py` could as well be a directory called `code` with a file `code/__init__.py`, where the above rules apply recursively.

Now, when you are inside of a python package (a directory that contains an `__init__.py` file), you should never import your own package. This might end up in circular imports. For you, it was working so far since you did not include all your files into the package (in the `__init__.py` file you did not import objects from some files).

There is an exception: when you run stand-alone code, e.g. scripts that are not inside of any `__init__.py` file, you can use either relative imports (`from .. import`) or absolute (`from package import`). In that case, I usually prefer the absolute import since then you can just copy that scripts anywhere else and run it -- this is usually good when you write example code.

[the remark this answers is not in the document]
Comment (akshay-raj-dhamija) [reply vast/losses/losses.py]: Thankyou!
